### ISSUE TITLE
AWS Lambda function invocation as data source (like CloudFormation custom resource)

### ISSUE THREAD

**Issue by aidansteele** [terraform-providers/terraform-provider-aws#2385]
Hiya,

CloudFormation has support for [custom resources](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html) in templates - functionality that is a limited subset of what Terraform supports in the form of providers' data sources. One of the supported receivers for CFN custom resources is a Lambda function. When specified, CloudFormation _invokes_ the Lambda function and passes properties to the function as a request body. The response body is then available to the rest of the template as attributes.

It would be great to have similar functionality in Terraform. We can almost do this using the [External Data Source](https://www.terraform.io/docs/providers/external/data_source.html) escape hatch, by doing something like:

```
data "external" "example" {
  program = ["bash", "-c", "xargs -0 -I '{}' aws lambda invoke --function-name LAMBDA_ARN --payload '{}' /dev/stdout"]

  query = {
    id = "abc123"
  }
}
```

But this is fiddly to type, requires the AWS CLI to be installed and credentials are a bit of a pain to manage. It would be great if this had first-class support.

What do you think? If there's general approval of the idea - or it needs some tweaks - I'd be happy to submit a PR.

**Comment by bflad** [comment]
A new [`aws_lambda_invocation` data source](https://www.terraform.io/docs/providers/aws/d/lambda_invocation.html) was previously released in [version 1.19.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1190-may-16-2018) and has been available in all releases since. Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by benishak** [comment]
@bflad this still does not do what "Custom Resources" do ...
does Terraform run aws_lambda_invocation only at creation ?
does Terraform pass Action Type to the Lambda like "CREATE", "UPDATE", "DELETE" so the lambda knows what to do? or can I use something like `${terraform.actiontype` or any other variable that holds the current terraform action ("apply", "destroy" ...) ?
